src/AutocompleteBadges.soy

```diff
@@ -0,0 +1,22 @@
+{namespace AutocompleteBadges}
+
+/**
+ * This renders the component's whole content.
+ * @param? badges
+ * @param? elementClasses
+ */
+{template .render}
+
+	<div class="autocomplete-badges {$elementClasses ?: ''}">
+		<textarea ref="input" autocomplete="off" class="autocomplete-field form-control" wrap="off"></textarea>
```

Comment:
So, let me explain the idea I've mentioned of how to get rid of those width calculations in the js file.

Basically, instead of changing the input field's padding every time we add a badge, what most components like this do is to position the input field after the badges list. That way the cursor will always be in the correct position.

The only immediate problem with this approach is that we want the user to think that the badges are **inside** the input, not next to it, right? That can be done by removing the borders of the input field entirely (with `border: none; box-shadow: none;` in the css file), and adding a border to the root div (the one with the **autocomplete-badges** class) instead (something like `border: 1px solid #ccc;`).

The only problem left after this is how to allow the input to fill the remaining space after the badge list. But I found some css that takes care of this too 😃

For all this to work you first need to move the **textarea** to be after the badge list, and also wrapped by a span, like this:

```html
<ul ref="listBadges" class="autocomplete-badges--list">...</ul>
<span class="autocomplete-field-wrapper">
  <textarea ref="input" autocomplete="off" class="autocomplete-field form-control" wrap="off"></textarea>
</span>
```

Then, in the css, you should have something like this:
```css
.autocomplete-badges {
	border: 1px solid #ccc;
	overflow: hidden;

	.autocomplete-field {
		border: none;
		box-shadow: none;
	}

	.autocomplete-field-wrapper {
		display: block;
		overflow: hidden;
	}

	&--list {
		float: left;
  }
}
```

Those are just the required css rules for this to work, you can keep the other rules you're already using like **padding**, **height** and **margin**. Just don't forget to remove the `position: relative` and `position: absolute` ones, we won't be needing those.

I could do these changes on my computer, but I thought they were really interesting and so wanted to share them with you instead of just changing things. Let me know if you are too busy with other things though, and I can do the changes myself :)